src/server/shell/token_authority.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mir/shell/token_authority.h"
+
+#include "mir/main_loop.h"
+#include "mir/time/alarm.h"
+#include "mir/log.h"
+
+#include <chrono>
+#include <cmath>
+#include <memory>
+#include <mutex>
+#include <uuid.h>
+
+namespace
+{
+/// Time in milliseconds that the compositor will wait before invalidating a token
+static auto constexpr token_invalidation_timeout = std::chrono::seconds(3);
+}
+
+namespace msh = mir::shell;
+
+msh::TokenAuthority::Token::Token(
+    std::string token, std::optional<RevocationListener> revocation_listener) :
+    token{token},
+    revocation_listener{revocation_listener}
+{
+}
+
+msh::TokenAuthority::Token::operator std::string() const
+{
+    return token;
+}
+
+bool msh::TokenAuthority::Token::operator==(Token const& token) const
+{
+    return this->token == token.token;
+}
+
+msh::TokenAuthority::TokenAuthority(std::shared_ptr<MainLoop>&& main_loop) :
+    main_loop{main_loop}
```

Comment:
We shouldn't have a constructor that accepts a transfer of ownership and tests that rely on it not transferring